### ISSUE TITLE
CSS module styling is removed too early on route changes

### ISSUE THREAD

**Issue by cmwall** [vercel/next.js#17464]
# Bug report

## Describe the bug

CSS module styling is removed immediately after clicking a `next/link`, instead of after the DOM is removed **on production builds**. This causes the components to have no styling at all during a page transition. **This issue does not happen on dev mode**.

I believe this is a bug with CSS modules specifically because components styled with `styled-jsx` don't have this problem.

Really would love to be able to use Sass via CSS modules here instead of re-writing the entire app I'm working on using `styled-jsx`. If Sass modules can't work in this scenario, I think I would be forced to use `styled-jsx`, which is not my preferred method of styling my components for this project. 

## To Reproduce

I have created repos, and deployed these repos to demonstrate the problem using [framer-motion](https://github.com/framer/motion) for page transitions. If you were to pull these repos and run them locally using `npm run dev`, you will see that the flash of unstyled content does not happen on any one of them **in dev mode**. However, on their deployed sites, you can see the flash of unstyled content with CSS modules and Sass modules.

### styled-jsx 

**Behavior**: correct, no flash of unstyled content
[Deployed site on Vercel](https://styled-jsx-leaving.vercel.app)
[Repo](https://github.com/cmwall/styled-jsx-leaving)

### CSS modules 

**Behavior**: buggy, there is a flash of unstyled content immediately after clicking the link
[Deployed site on Vercel](https://css-modules-leaving.vercel.app)
[Repo](https://github.com/cmwall/css-modules-leaving)

### Sass via CSS modules (additional)

**Behavior**: buggy, there is a flash of unstyled content immediately after clicking the link (same as CSS modules)
[Deployed site on Vercel](https://sass-modules-leaving.vercel.app)
[Repo](https://github.com/cmwall/sass-modules-leaving)

## Expected behavior

Styling for components that come from CSS modules should not be removed immediately on route changes, and instead, are removed when the markup is removed (the component unmounts?). The expected behavior is the behavior we can see on the `styled-jsx` deployment above. 

## System information

- OS: macOS
- Browser (if applies): N/A
- Version of Next.js: 9.5.3
- Version of Node.js: 12.14.1

**Comment by MihirGH** [comment]
I looked at this issue and this seems to be happening because of the behaviour implemented in `onCommit` method.

Inside `onCommit`, we are setting the media attribute to the style tags which are not desired too early. 

Exploring this further and will raise a PR for this in sometime

**Comment by MihirGH** [comment]
Essentially what I ended up finding was: we should never set `media` attribute to `x` on the `<style>` tags since they might still be in use even though the page that was using it is being replaced. To do this, we simply need to get rid of the code that is executing in the else block of [this loop](https://github.com/vercel/next.js/blob/canary/packages/next/client/index.tsx#L637-L639) in `onCommit`.

I don't know if this was done because of any specific reasons or not but this seems to solve the issue on my local after making these changes.

@lfades any thoughts?

**Comment by MihirGH** [comment]
Also, to make this work and to have the `<style>` tags on SSRed page, we might need to remove the `isInitialRender` check in `onStart` and `!isInitialRender` check in `onCommit`

**Comment by ccambournac** [comment]
Hi, experiencing the very same issue as reported.

**Comment by scriptify** [comment]
Experiencing the same issue. I also notice the `media` attribute being set to `x` on the relevant `style` tags. Would love to avoid the switch to styled-jsx because of this issue, because styled-jsx doesn't play well with framer motion: https://github.com/framer/motion/issues/231

I've applied this hack in the mean time:

```javascript
    const routeChangeComplete = () => {
      // Temporary fix to avoid flash of unstyled content
      // during route transitions. Keep an eye on this
      // issue and remove this code when resolved:
      // https://github.com/vercel/next.js/issues/17464

      const tempFix = () => {
        const allStyleElems = document.querySelectorAll('style[media="x"]');
        allStyleElems.forEach((elem) => {
          elem.removeAttribute("media");
        });
      };
      tempFix();
    };
   Router.events.on("routeChangeComplete", routeChangeComplete);
```

I am not sure what other unwished side effects this temporary fix could have, but it seems to work just well enough for my application.

**Comment by tommhuth** [comment]
I'm having the same issue, but @scriptify's fix does not work for me, any ETA for a fix for this? @MihirGH

**Comment by MihirGH** [comment]
@tommhuth It is actually a quick fix but I am not sure if that's how it is supposed to be solved or not.

**Comment by scriptify** [comment]
@MihirGH That code is just a temporary workaround, it also possible that Next.js is setting the media attribute slightly after those events fire, maybe that's the reason why it's not working for @tommhuth. @tommhuth are you sure that in your case the problem is also related to the `media` attribute?

**Comment by tommhuth** [comment]
Having looked into this, I can't see anywhere that  media attribute is being changed, but I do see that the relevant `link` has its `rel` set from `stylesheet` to `preload`. If I manually revert it to `stylesheet` the appropriate styles are back, but when I do this in a `routeChangeComplete`/`routeChangeComplete` event callback there is a flash of unstyled content so those events are probably not accurate enough to serve as a quick fix. Also, I don't think I have any way of knowing exactly what style `link` is needed, so I end up setting all links with `css` files back to `stylesheet` which might be unnecessary.  @scriptify @MihirGH 

Off the top of my head, the only way I could get around this is by moving all style modules into non-module SCSS, but that is quite a significant workaround.

**Comment by scriptify** [comment]
Hmm that's strange, seems like a severe issue to me, and moving everything into non modular CSS seems like a very tedious task, just to make it modular again when it's fixed (besides the whole disadvantages global CSS has). Are we the only ones using CSS Modules + non-instant route changes? 🤷‍♂️ Doesn't seem too exotic to me

**Comment by Limekiller** [comment]
@scriptify I was running into this issue as well; your hack worked for me for the time being, so thanks for that.

**Comment by fredcorr** [comment]
@scriptify I bumped into the same issue. could you help me understand where I would add the fix you wrote? Thanks for your help as well

**Comment by Limekiller** [comment]
@fredcorr I put the code in _app.js along with, of course, `import Router from 'next/router'`

**Comment by scriptify** [comment]
@fredcorr Yea that code is a bit out of context, as @Limekiller mentioned, the best place to put it is `_app.tsx` /  `_app.jsx`. I updated my comment accordingly to clear that up.

**Comment by fredcorr** [comment]
@Limekiller @scriptify thanks guys that fixed the issue partially, on the first-page transition the issue still occurs. Are you guys using the getStaticProps or GgetServerSideProps? Could that maybe affect it? @MihirGH Any updated on how long will it take to fix this?

**Comment by scriptify** [comment]
@fredcorr I'm using `getStaticProps` for all my pages for this particular project, but I don't think that should affect the route transition behaviour, just speculating ofc. Would be nice to get some hints from the Next.js team 🤗

**Comment by tommhuth** [comment]
For anyone needing a dirty quick fix for this one until a fix is released, simply importing the modules whose style is needed in `_app.js` solved it for me without requiring any other refactoring.

**Comment by OlivierFortier** [comment]
I am having the exact same problem, and @scriptify 's fix seemed to have resolved the issue for me in general.

However it still has the same issue but only on the first page transition (regardless of which page) , afterwards every page transition to any page in any order seems fine. I guess that will have to do until an official fix comes.

**Comment by controversial** [comment]
@Timer I can confirm that #19125 didn’t fix this, as I’m still seeing this issue on 10.0.3 :(

**Comment by timneutkens** [comment]
> @Timer I can confirm that #19125 didn’t fix this, as I’m still seeing this issue on 10.0.3 :(

Would be useful if you provide a reproduction for your particular case.

**Comment by cmwall** [comment]
@timneutkens Would it help if I updated the original demonstration repos to 10.0.3?

**Comment by nayarahilton** [comment]
@scriptify thanks for the partial solution! Some thoughts in how to fix that for the first-page transition?

**Comment by controversial** [comment]
> Would be useful if you provide a reproduction for your particular case.

I think the original reproduction from @cmwall attached to this issue represents my use case, but I can deploy another copy with version 10.0.3 if that helps

**Comment by ShuPink** [comment]
Hiya, I've also had this problem on 10.0.3, using react-transition-group and css modules.

@timneutkens I've tried my hand at a reproduction [here](https://next-17464.vercel.app/).
If you go from index-> page 1, you'll see the styles be removed from the index page for second.
If you go from page 1 -> page 2 there won't be an issue because the same stylesheet is used.

If you run this in dev the transitions work as expected.

The repo is [here](https://github.com/ShuPink/next-17464)
Um. Let me know if the demo isn't clear.

Here's a screenshot of the index page when it exits and it's styles are removed
![during transition](https://user-images.githubusercontent.com/60080324/102432948-6faa7180-404e-11eb-9511-2ce6a3944cfa.PNG)

Vs. pre-transition
![pre-transition](https://user-images.githubusercontent.com/60080324/102433000-72a56200-404e-11eb-86e3-2f9b291a5549.PNG)

**Comment by mrgnou** [comment]
This issue is quite old and breaks every page transitioning website since nextjs 9.3. Shouldn't it be marked as critical and moved to the next milestone?

**Comment by stefanmaric** [comment]
I've encounter this issue as well and tried @scriptify 's [suggested woraround](https://github.com/vercel/next.js/issues/17464#issuecomment-711330281) but—as other have mentioned—it didn't seem to really help with the first navigation.

After looking deeper, I realized this is a hard problem to solve for the framework. I don't really know how pre-v10 versions of Next.js worked in this regard since the project I'm working on started with the v10, but I can assume any page transition solution relied on a bug or an unoptimized behavior.

When the page changes you want to remove styles (and maybe other kind of resources) from the DOM to 1) prevent style clashes (in the case of global CSS) and to 2) prevent memory leaks as the CSSOM would otherwise increase on every page navigation.

Next.js handles this well.

The problem with page transitions arises from the fact that any solution (being it with Framer motion, ReactTransitionGroup, ReactFlipToolkit, etc) relies on holding to the previous render's element tree (its `children`) until the animation has finished.

Next.js cannot know when such animation finishes, instead, it has to unlink styles from the DOM as soon as the current (old) page `Component` is replaced by the next (new) page `Component`.

I don't think here's any "fix" Next.js can apply for this. The only solution I can think of is for Next.js to provide a new set of APIs to hook to and manipulate its resource management system on the client-side. Think of providing callbacks on the `_app` interface or making the `Route` events an async middleware that allows one to delay the next step in the route change process.

# tl;dr: a workaround

I've come up with a workaround that 1) seems to be working even for the first navigation and 2) cleans up the DOM after the page transition has finished.

There's 2 variations depending on the animation system you use. Note you would use **one** or the **other** depending on your case, **not both**.

## Spring-based

The first one is for spring-based animations which provide a completion callback. Place this hook anywhere you prefer:

```typescript
// utils/useTransitionFix.ts

import Router from 'next/router'
import { useCallback, useEffect, useRef } from 'react'

type Cleanup = () => void

export const useTransitionFix = (): Cleanup => {
  const cleanupRef = useRef<Cleanup>(() => {})

  useEffect(() => {
    const changeListener = () => {
      // Create a clone of every <style> and <link> that currently affects the page. It doesn't
      // matter if Next.js is going to remove them or not since we are going to remove the copies
      // ourselves later on when the transition finishes.
      const nodes = document.querySelectorAll('link[rel=stylesheet], style:not([media=x])')
      const copies = [...nodes].map((el) => el.cloneNode(true) as HTMLElement)

      console.log(copies)

      for (let copy of copies) {
        // Remove Next.js' data attributes so the copies are not removed from the DOM in the route
        // change process.
        copy.removeAttribute('data-n-p')
        copy.removeAttribute('data-n-href')

        // Add duplicated nodes to the DOM.
        document.head.appendChild(copy)
      }

      cleanupRef.current = () => {
        for (let copy of copies) {
          // Remove previous page's styles after the transition has finalized.
          document.head.removeChild(copy)
        }
      }
    }

    Router.events.on('beforeHistoryChange', changeListener)

    return () => {
      Router.events.off('beforeHistoryChange', changeListener)
      cleanupRef.current()
    }
  }, [])

  // Return an fixed reference function that calls the internal cleanup reference.
  return useCallback(() => {
    cleanupRef.current()
  }, [])
}

```

Then you can use it in your `app`, e.g.:

```typescript
// pages/_app.ts

import { AnimatePresence, motion } from 'framer-motion'
import type { AppProps } from 'next/app'

import { useTransitionFix } from '../utils/useTransitionFix'

const PAGE_VARIANTS = {
  initial: {
    opacity: 0,
  },
  animate: {
    opacity: 1,
  },
  exit: {
    opacity: 0,
  },
}

function App({ Component, pageProps, router }: AppProps): React.ReactElement {
  const transitionCallback = useTransitionFix()

  return (
      <AnimatePresence exitBeforeEnter onExitComplete={transitionCallback}>
        <motion.div
          key={router.route}
          initial="initial"
          animate="animate"
          exit="exit"
          variants={PAGE_VARIANTS}
        >
          <Component {...pageProps} />
        </motion.div>
      </AnimatePresence>
  )
}

export default App

```
## Timeout-based

The second one is for solutions that use fixed-duration transitions:

```typescript
// utils/fixTimeoutTransition.ts

import Router from 'next/router'

export const fixTimeoutTransition = (timeout: number): void => {
  Router.events.on('beforeHistoryChange', () => {
    // Create a clone of every <style> and <link> that currently affects the page. It doesn't matter
    // if Next.js is going to remove them or not since we are going to remove the copies ourselves
    // later on when the transition finishes.
    const nodes = document.querySelectorAll('link[rel=stylesheet], style:not([media=x])')
    const copies = [...nodes].map((el) => el.cloneNode(true) as HTMLElement)

    for (let copy of copies) {
      // Remove Next.js' data attributes so the copies are not removed from the DOM in the route
      // change process.
      copy.removeAttribute('data-n-p')
      copy.removeAttribute('data-n-href')

      // Add duplicated nodes to the DOM.
      document.head.appendChild(copy)
    }

    const handler = () => {
      // Emulate a `.once` method using `.on` and `.off`
      Router.events.off('routeChangeComplete', handler)

      window.setTimeout(() => {
        for (let copy of copies) {
          // Remove previous page's styles after the transition has finalized.
          document.head.removeChild(copy)
        }
      }, timeout)
    }

    Router.events.on('routeChangeComplete', handler)
  })
}

```

Which you would use **outside** of of your `app` component:

```typescript
// pages/_app.ts
import { CSSTransition, TransitionGroup } from "react-transition-group"
import type { AppProps } from 'next/app'

import { fixTimeoutTransition } from '../utils/useTransitionFix'

import '../styles/globals.css'

const TRANSITION_DURATION = 500

fixTimeoutTransition(TRANSITION_DURATION)

function App({ Component, pageProps, router }: AppProps): React.ReactElement {
  return (
    <TransitionGroup>
      <CSSTransition
        classNames="app-transition-wrapper"
        enter
        exit
        key={router.asPath}
        timeout={TRANSITION_DURATION}
        unmountOnExit
      >
        <Component {...pageProps} />
      </CSSTransition>
    </TransitionGroup>
  )
}

export default App

```

So far I haven't found issues with these. Hope it helps you all.

**Comment by wushaobo** [comment]
It is quite a scary news for my team, since we just started to switch from styled-components to css modules a week ago.
Has it been fixed since 10.0.4? Is there a rough estimation time of release in your plan? Please kindly update. Thanks.

**Comment by controversial** [comment]
This is the [most upvoted open bug on this repo](https://github.com/vercel/next.js/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc+label%3A%22kind%3A+bug%22) right now

**Comment by ccambournac** [comment]
@stefanmaric 
Thanks for the workaround! 
In my case, I had to use the `routeChangeStart` event for it to work.

**Comment by controversial** [comment]
Worth noting that in addition to styles being removed, **the page’s scroll position also jumps to the top** at the beginning of the transition.

**Comment by ccambournac** [comment]
Still experiencing small blinks on first navigation...
https://www.loom.com/share/ee94e6065c2b4ba488601176c39323d4

(But no scroll position jump @controversial)

**Comment by ClayCooperLA** [comment]
I fixed this by importing the component that is leaving into the `_app.js` file.

In my particular case, we only have one component that exposes the issue, so simply importing that component into app.js, not even rendering it, fixes the issue. This will probably work for other cases, you would just import any component that loses its styles during a transition into the `_app.js` file. Obviously you will lose some code splitting magic, but in our case, we feed optional data into the components so it's really just an empty version of the component that is being imported with its styles.

**Comment by Timer** [comment]
@stefanmaric has explained this situation quite well.

tl;dr Next.js handles CSS removal and injection correctly when no libraries "hold onto" previous rendering trees.

I've included the relevant bit here:

> The problem with page transitions arises from the fact that any solution (being it with Framer motion, ReactTransitionGroup, ReactFlipToolkit, etc) relies on holding to the previous render's element tree (its children) until the animation has finished.
>
> Next.js cannot know when such animation finishes, instead, it has to unlink styles from the DOM as soon as the current (old) page Component is replaced by the next (new) page Component.

Next.js cannot simultaneously allow rendering with both sets of styles (from two separate pages) on screen, as this may result in styles conflicts which would trigger reflows or the incorrect styling of your application.

FWIW, this works with `styled-jsx` because the styles are defined as part of the render-tree instead of as a module side-effect.

It'd be great to see one of these libraries get the computed styles of each element and apply them as inline instead of thinking they'd never be removed from the page.

---

I've marked this as good first issue for someone to investigate further, as it'd probably be good to have a story around this.

**Comment by Gr0m** [comment]
Why was it working with 9.5.2? Don't want to stick with this version forever.

**Comment by Timer** [comment]
9.5.2 and earlier had numerous bugs with styling conflicts that we fixed (not sure this is the exact version, but going off what you said), as mentioned above:

> Next.js cannot simultaneously allow rendering with both sets of styles (from two separate pages) on screen, as this may result in styles conflicts which would trigger reflows or the incorrect styling of your application.

To expand on this, we used to treat styles across pages as "append-only" which broke applications in all sorts of ways.
By removing old styles, we closed over half a dozen issues with tons of upvotes. Unfortunately, fixing this issue had the side effect of revealing a bug/lack-of-feature in these transition libraries that do not snapshot styles.

**Comment by Gr0m** [comment]
That's unfortunate. But what you say makes sense.

So styled-jsx will still work. But what about Styled Components? Or will it fail the same way as CSS Modules do? Or maybe it will be better to use one of @stefanmaric solutions?

I can leave my old projects on 9.5.2, but need to have a working solution for future projects.

I really like Next.js and don't want to use anything else (I have tried - didn't like :) )

**Comment by Timer** [comment]
Styled Components and Styled-JSX should both work fine with transition libraries. We're open to someone doing research into what it'd take to make styles append-only again, or if it's even possible without conflicts.

**Comment by claudiobonfati** [comment]
The @stefanmaric [suggestion](https://github.com/vercel/next.js/issues/17464#issuecomment-751267740) is a good fix for the most cases, but it causes a serious memory leak when a dynamic page change within the same route structure. 
For example, if you have a route `/profile/[username]`, and you switch pages within the same route, like from `/profile/vader` to `/profile/yoda` the Cleanup function will not be triggered, which will clone the CSS nodes from DOM over and over again (**4** nodes will be **8**, **8** will be **16**, **16** to **32**, **32** to **64**, **128**, **256**, **512**...)

**Comment by claus** [comment]
Here's a hack that basically restores the Next 9.5 behavior in Next 10 (all page styles are preserved).

```javascript
useEffect(() => {
    document
        .querySelector('head > link[rel="stylesheet"][data-n-p]')
        ?.removeAttribute('data-n-p');
    const mutationHandler = mutations => {
        mutations.forEach(mutation => {
            const { type, target } = mutation;
            if (type === 'attributes' && target.nodeName === 'STYLE') {
                const newValue = target.getAttribute('media');
                if (newValue === 'x') {
                    target.removeAttribute('media');
                }
            }
        });
    };
    const observer = new MutationObserver(mutationHandler);
    observer.observe(document.head, {
        subtree: true,
        childList: true,
        attributeFilter: ['media'],
    });
    return () => {
        observer.disconnect();
    };
}, []);
```

On mount, it removes the `data-n-p` attribute of the `<link rel="stylesheet" data-n-p>` so that Next doesn't remove the element anymore. It also listens to changes in the `media` attributes of the page `<style>`s that are added when navigating to a new page. When a `media="x"` is added, it removes it again straight away.

Pro: Next won't remove any styles anymore
Con: Next won't remove any styles anymore (watch out for possible conflicts)

**Comment by timohofmeijer** [comment]
In response to [stefanmaric’s fix](https://github.com/vercel/next.js/issues/17464#issuecomment-751267740):
> So far I haven't found issues with these. Hope it helps you all.

Running this fix in `NODE_ENV` `'development'` caused serious issues for me (the react-native ipad app wrapping my next app crashing consistently), so I had to tweak it to only run in `'production'`. Perhaps this info can save someone some time.

**Comment by abdulrauf11** [comment]
Has this issue been fixed yet? 😢

**Comment by satazor** [comment]
Unfortunately not, at least in the latest released Next.js version.

I’ve made a solution based @claus approach that has been working well for us: https://github.com/moxystudio/next-with-moxy/blob/master/www/app/use-fouc-fix.js

**Comment by mattvb91** [comment]
Not sure if this is related to the same thing but im currently experiencing css not being rendered on route changes on the next page if the css is imported for the component from node_modules. When I hard refresh for SSR it works fine

**Comment by adedaniel** [comment]
I had been battling with this issue for weeks now, occasionally when I change the route using `Router.push`, the css in the new page breaks.

I fixed this by using [scriptify's answer](https://github.com/vercel/next.js/issues/17464#issuecomment-711330281). 
It seems to have fixed the problem. 🙂

**Comment by SalmanK81099** [comment]
> Experiencing the same issue. I also noticed the `media` attribute being set to `x` on the relevant `style` tags. Would love to avoid the switch to styled-jsx, as it doesn't play well with framer motion: [framer/motion#231](https://github.com/framer/motion/issues/231)
> 
> I've applied this hack in the mean time:
> 
> ```js
>     // Add that code to _app.tsx / _app.jsx
> 
>     import Router from "next/router";
> 
>     const routeChange = () => {
>       // Temporary fix to avoid flash of unstyled content
>       // during route transitions. Keep an eye on this
>       // issue and remove this code when resolved:
>       // https://github.com/vercel/next.js/issues/17464
> 
>       const tempFix = () => {
>         const allStyleElems = document.querySelectorAll('style[media="x"]');
>         allStyleElems.forEach((elem) => {
>           elem.removeAttribute("media");
>         });
>       };
>       tempFix();
>     };
> 
>    Router.events.on("routeChangeComplete", routeChange );
>    Router.events.on("routeChangeStart", routeChange );
> ```
> 
> I am not sure what other unwished side effects this temporary fix could have, but it seems to work just well enough for my application.

Hi, this code works perfectly but not on the first load I managed to get it working by adding the following in the app component.
React.useEffect(() => {
    router.push(router.pathname, { shallow: true });
  }, []);

**Comment by jacobbroughton** [comment]
> React.useEffect(() => {
>     router.push(router.pathname);
>   }, []);
> ```

This worked great for me but refreshing the page on a dynamic route gave me an error to the effect of 'Link component interpolation error', adding the query value and passing all current queries to the router.push inside of the useeffect fixed it for me. Hopefully this'll come in handy for someone!

    router.push({
      pathname: router.pathname,
      query: {...router.query}
    });

**Comment by SalmanK81099** [comment]
> > React.useEffect(() => {
> > router.push(router.pathname);
> > }, []);
> 
> This worked great for me but refreshing the page on a dynamic route gave me an error to the effect of 'Link component interpolation error', adding the query value and passing all current queries to the router.push inside of the useeffect fixed it for me. Hopefully this'll come in handy for someone!
> 
> ```
> router.push({
>   pathname: router.pathname,
>   query: {...router.query}
> });
> ```

I got to face the same problem but don't know why this solution did not work for me, Since on the first render of a dynamic link the router.pathname returns /[slug] or any param you passed instead of the actual param value so it will give you an interpolation error to fix it simply replace router.pathname by javascripts window.location.pathname which gives us the exact path and thus solves the problem without requiring to pass any queries.

`router.push({
      pathname: window.location.pathname,
    });`

This will work with all types of links

**Comment by yosefbeder** [comment]
> 
> 
> Here's a hack that basically restores the Next 9.5 behavior in Next 10 (all page styles are preserved).
> 
> ```js
> useEffect(() => {
>     Array.from(
>         document.querySelectorAll('head > link[rel="stylesheet"][data-n-p]')
>     ).forEach(node => {
>         node.removeAttribute('data-n-p');
>     });
>     const mutationHandler = mutations => {
>         mutations.forEach(({ target }) => {
>             if (target.nodeName === 'STYLE') {
>                 if (target.getAttribute('media') === 'x') {
>                     target.removeAttribute('media');
>                 }
>             }
>         });
>     };
>     const observer = new MutationObserver(mutationHandler);
>     observer.observe(document.head, {
>         subtree: true,
>         attributeFilter: ['media'],
>     });
>     return () => {
>         observer.disconnect();
>     };
> }, []);
> ```
> 
> On mount, it removes the `data-n-p` attribute of the `<link rel="stylesheet" data-n-p>` elements so that Next doesn't remove those elements anymore. It also listens to changes in the `media` attributes of the page `<style>`s that are added when navigating to a new page. When a `media="x"` is added, it removes it again straight away.
> 
> Pro: Next won't remove any styles anymore
> Con: Next won't remove any styles anymore (watch out for possible conflicts)

Thanks that helped me a lot

**Comment by natashajaved** [comment]
Hey, I am adding a css file using <link/> in a specific page, I'm running into a flashing issue, where the page appears without the new css and then styling appears, It only happens when I'm navigating to the page from some other page and not when I directly land on the page, tried @scriptify solution, it doesn't change anything.
Appreciate any help.
Adding link in head
`   <Head>
            <link 
                rel="stylesheet" href="/style-quran.css" />
        </Head>`

**Comment by tomasztunik** [comment]
Also stumbled on the same issue when working with framer-motion. The solution worked I only had to replace `router.push(router.pathname)` with `router.push(router.asPath)` otherwise it would throw errors on parametrised URLs when refreshing giving following error:

```
Error: The provided `href` (/page-name/[param]) value is missing query values (slug) to be interpolated properly. Read more: https://nextjs.org/docs/messages/href-interpolation-failed
```

this would happen because `router.pathname` returns route pattern not the actual path with filled pathname parameters values.

**Comment by thibault-barrat** [comment]
> Also stumbled on the same issue when working with framer-motion. The solution worked I only had to replace `router.push(router.pathname)` with `router.push(router.asPath)` otherwise it would throw errors on parametrised URLs when refreshing giving following error:
> 
> ```
> Error: The provided `href` (/page-name/[param]) value is missing query values (param) to be interpolated properly. 
> Read more: https://nextjs.org/docs/messages/href-interpolation-failed
> ```
> 
> this would happen because `router.pathname` returns route pattern not the actual path with filled pathname parameters values. Ie. `/page/[slug]` instead of `/page/slug-uid`

This works for me but caused a bug when trying to go to an url which not exists and which displays the 404 page. I had an infinite refresh loop because the not found url is pushed by the router. I solved it with the following code : 
```jsx
useEffect(() => {
    if (router.pathname === '/404') {
      router.push(router.pathname);
    } else {
      router.push(router.asPath);
    }
  }, []);
```

**Comment by balazsorban44** [comment]
This is still reproducible using `12.0.11-canary.19` and the example from #33785

**Comment by panteliselef** [comment]
I am having the same issue, with styles being removed the moment I am clicking on a <Link />. Current set up is 
``` 
"framer-motion": "6.2.8",
"next": "12.1.0", 
"react": "17.0.2",    
``` 
I used the following to fix it and for now it works like a charm. It is from code written above, i just wrapped it in a hook

```tsx
import Router from 'next/router';
import { useEffect } from 'react';

const routeChange = () => {
    const tempFix = () => {
        const allStyleElems = document.querySelectorAll('style[media="x"]');
        allStyleElems.forEach((elem) => {
            elem.removeAttribute('media');
        });
    };
    tempFix();
};

export const useTransitionFix = (): void => {
    useEffect(() => {
        Router.events.on('routeChangeComplete', routeChange);
        Router.events.on('routeChangeStart', routeChange);

        return () => {
            Router.events.off('routeChangeComplete', routeChange);
            Router.events.off('routeChangeStart', routeChange);
        };
    }, []);

    useEffect(() => {
        Router.router?.push(Router.router?.pathname);
    }, []);
};

```

**Comment by dhruvindsd-dev** [comment]
> I am having the same issue, with styles being removed the moment I am clicking on a . Current set up is
> 
> ```
> "framer-motion": "6.2.8",
> "next": "12.1.0", 
> "react": "17.0.2",    
> ```
> 
> I used the following to fix it and for now it works like a charm. It is from code written above, i just wrapped it in a hook
> 
> ```tsx
> import Router from 'next/router';
> import { useEffect } from 'react';
> 
> const routeChange = () => {
>     const tempFix = () => {
>         const allStyleElems = document.querySelectorAll('style[media="x"]');
>         allStyleElems.forEach((elem) => {
>             elem.removeAttribute('media');
>         });
>     };
>     tempFix();
> };
> 
> export const useTransitionFix = (): void => {
>     useEffect(() => {
>         Router.events.on('routeChangeComplete', routeChange);
>         Router.events.on('routeChangeStart', routeChange);
> 
>         return () => {
>             Router.events.off('routeChangeComplete', routeChange);
>             Router.events.off('routeChangeStart', routeChange);
>         };
>     }, []);
> 
>     useEffect(() => {
>         Router.router?.push(Router.router?.pathname);
>     }, []);
> };
> ```

Great temporary fix, but we can make this better by adding a `setTimeout`. This will make sure that the useless styles are removed after the component is done unmounting. This prevents crashes in large applications. 
```tsx
const handleRouteChange = () => {
	const elements = document.querySelectorAll('style[media="x"]');
	elements.forEach((elem) => elem.removeAttribute("media"));
	setTimeout(() => {
		elements.forEach((elem) => elem.remove());
	}, OPACITY_EXIT_DURATION * 1000);
};

Router.events.on("routeChangeStart", handleRouteChange);
Router.events.on("routeChangeComplete", handleRouteChange);
```

**Comment by BenMasel** [comment]
Hi guys! I'm very new to all this and kept thinking I had done something wrong. Glad to know I was not the problem! Speaking of which how do I implement that temporary fix into my own code?

**Comment by clieee** [comment]
@BenMasel put this code:

```
import Router from 'next/router';
import { useEffect } from 'react';

export const OPACITY_EXIT_DURATION = 1;

const routeChange = () => {
    const tempFix = () => {
        const elements = document.querySelectorAll('style[media="x"]');
        elements.forEach((elem) => elem.removeAttribute('media'));
        setTimeout(() => {
            elements.forEach((elem) => elem.remove());
        }, OPACITY_EXIT_DURATION * 1000);
    };
    tempFix();
};

export const useTransitionFix = () => {
    useEffect(() => {
        Router.events.on('routeChangeComplete', routeChange);
        Router.events.on('routeChangeStart', routeChange);

        return () => {
            Router.events.off('routeChangeComplete', routeChange);
            Router.events.off('routeChangeStart', routeChange);
        };
    }, []);

    useEffect(() => {
        Router.router?.push(Router.router?.pathname);
    }, []);
};
```

in a useTransitionFix.js and then in your _app.js you can just use ```useTransitionFix()``` so the code is invoked.

**Comment by lassejaco** [comment]
Just throwing in my voice here - this was a huge pain to fix
Nextjs is really awesome but I'm exceedingly running into quite critical issues :-/

**Comment by vietanhtran16** [comment]
> @BenMasel put this code:
> 
> ```
> import Router from 'next/router';
> import { useEffect } from 'react';
> 
> export const OPACITY_EXIT_DURATION = 1;
> 
> const routeChange = () => {
>     const tempFix = () => {
>         const elements = document.querySelectorAll('style[media="x"]');
>         elements.forEach((elem) => elem.removeAttribute('media'));
>         setTimeout(() => {
>             elements.forEach((elem) => elem.remove());
>         }, OPACITY_EXIT_DURATION * 1000);
>     };
>     tempFix();
> };
> 
> export const useTransitionFix = () => {
>     useEffect(() => {
>         Router.events.on('routeChangeComplete', routeChange);
>         Router.events.on('routeChangeStart', routeChange);
> 
>         return () => {
>             Router.events.off('routeChangeComplete', routeChange);
>             Router.events.off('routeChangeStart', routeChange);
>         };
>     }, []);
> 
>     useEffect(() => {
>         Router.router?.push(Router.router?.pathname);
>     }, []);
> };
> ```
> 
> in a useTransitionFix.js and then in your _app.js you can just use `useTransitionFix()` so the code is invoked.

could someone explain why we need this router.push?

```
useEffect(() => {
        Router.router?.push(Router.router?.pathname);
    }, []);
```

**Comment by wilbsy** [comment]
Just a caveat that the router.push() portion of the fix detailed above [comment 949620547](https://github.com/vercel/next.js/issues/17464#issuecomment-949620547) will cause a missing query error on dynamic routes [#33243](https://github.com/vercel/next.js/discussions/33243).  Hope this can save someone some headache.

**Comment by wilbsy** [comment]
> > @BenMasel put this code:
> > ```
> > import Router from 'next/router';
> > import { useEffect } from 'react';
> > 
> > export const OPACITY_EXIT_DURATION = 1;
> > 
> > const routeChange = () => {
> >     const tempFix = () => {
> >         const elements = document.querySelectorAll('style[media="x"]');
> >         elements.forEach((elem) => elem.removeAttribute('media'));
> >         setTimeout(() => {
> >             elements.forEach((elem) => elem.remove());
> >         }, OPACITY_EXIT_DURATION * 1000);
> >     };
> >     tempFix();
> > };
> > 
> > export const useTransitionFix = () => {
> >     useEffect(() => {
> >         Router.events.on('routeChangeComplete', routeChange);
> >         Router.events.on('routeChangeStart', routeChange);
> > 
> >         return () => {
> >             Router.events.off('routeChangeComplete', routeChange);
> >             Router.events.off('routeChangeStart', routeChange);
> >         };
> >     }, []);
> > 
> >     useEffect(() => {
> >         Router.router?.push(Router.router?.pathname);
> >     }, []);
> > };
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > in a useTransitionFix.js and then in your _app.js you can just use `useTransitionFix()` so the code is invoked.
> 
> could someone explain why we need this router.push?
> 
> ```
> useEffect(() => {
>         Router.router?.push(Router.router?.pathname);
>     }, []);
> ```

The useEffect implements fix for initial load.

**Comment by wilbsy** [comment]
If using the hook @vietanhtran16 posted above, use a [URL object](https://nextjs.org/docs/api-reference/next/router#with-url-object) in the router.push() in order to avoid dynamic route (i.e. `dir/[slug]`) errors:
```js
useEffect(() => {
    const pathname = Router.router?.pathname
    const query = Router.router?.query
    Router.router?.push({ pathname, query })
  }, [])
```
That being said, I'm surprised this is necessary... seems so wrong.  Is this issue being fixed?

**Comment by panteliselef** [comment]
@wilbsy I have not tested this case, as it was not needed in the project work on. But I will use that in my snippet as well. The issue still seems to be there where running on production for `"next": "~12.1.5"`

**Comment by chris-schra** [comment]
above fixes didn't work for us, but this really hacky one did:

```javascript
    useEffect(() => {
        const elements = document.querySelectorAll('head link[rel=stylesheet]');
        for (const el of elements) {
            if (el.hasAttribute('data-n-p')) {
                el.removeAttribute('data-n-p');
            }
        }
    }, []);
```

**Comment by wilbsy** [comment]
> If using the hook @vietanhtran16 posted above, use a [URL object](https://nextjs.org/docs/api-reference/next/router#with-url-object) in the router.push() in order to avoid dynamic route (i.e. `dir/[slug]`) errors:
> 
> ```js
> useEffect(() => {
>     const pathname = Router.router?.pathname
>     const query = Router.router?.query
>     Router.router?.push({ pathname, query })
>   }, [])
> ```
> 
> That being said, I'm surprised this is necessary... seems so wrong. Is this issue being fixed?

And another issue with [this approach](https://github.com/vercel/next.js/issues/17464#issuecomment-1100500663)...
If you're using query params, this will delete them (i.e. `?QUERY_KEY=value&QUERY_KEY_2=value2` will not be transferred to the the `router.push()`).  To solve for this, either use:
```js
Router.router?.push(asPath)
```
or
```js
Router.router?.push({ pathname, query }, asPath)
```
Not sure yet if there's an advantage to one over the other.  

And I still can't believe this is necessary...  NUTS!

**Comment by ugglr** [comment]
Basically using page transitions on next with framer is not usable today. I've found good solutions using react-transition-group instead. https://reactcommunity.org/react-transition-group/

**Comment by JoshuaHolloway** [comment]
I am experiencing this issue of the styling failing upon page transition using React Transition Group.  The moment the page transition begins the styling from CSS-Modules fails and the user sees the elements un-styled.  Am I understanding this correctly that if I simply change all the styles to use non-CSS-modules then I should not experience this problem?

**Comment by panteliselef** [comment]
There is no official fix yet. Changing the way you use styling, for example to go from CSS modules to pure css or styled-components, is not always easy. Especially if your codebase includes a lot of components, you will have a lot of side-effects. The samples above solve the issues tho.

**Comment by JoshuaHolloway** [comment]
@panteliselef which sample?  Does anyone have a repo of the current temporary fix as a minimal demo?  That sure would be cool.  I am somewhat new to Next.js and that really would help me (and certainly lots of other people) out.

**Comment by panteliselef** [comment]
Check this: https://github.com/vercel/next.js/issues/17464#issuecomment-1080337949
it is the code from above

**Comment by panteliselef** [comment]
> > If using the hook @vietanhtran16 posted above, use a [URL object](https://nextjs.org/docs/api-reference/next/router#with-url-object) in the router.push() in order to avoid dynamic route (i.e. `dir/[slug]`) errors:
> > ```js
> > useEffect(() => {
> >     const pathname = Router.router?.pathname
> >     const query = Router.router?.query
> >     Router.router?.push({ pathname, query })
> >   }, [])
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > That being said, I'm surprised this is necessary... seems so wrong. Is this issue being fixed?
> 
> And another issue with [this approach](https://github.com/vercel/next.js/issues/17464#issuecomment-1100500663)... If you're using query params, this will delete them (i.e. `?QUERY_KEY=value&QUERY_KEY_2=value2` will not be transferred to the the `router.push()`). To solve for this, either use:
> 
> ```js
> Router.router?.push(asPath)
> ```
> 
> or
> 
> ```js
> Router.router?.push({ pathname, query }, asPath)
> ```
> 
> Not sure yet if there's an advantage to one over the other.
> 
> And I wonder if `router.replace()` would be better?
> 
> And I still can't believe this is necessary... NUTS!

When you are saying  `this will delete them` do you they will not be accessible from any hook or just removes them from the visible URL?

**Comment by wilbsy** [comment]
@JoshuaHolloway 
in __app.js_...
```jsx
import { usePageTransitionFix } from '@/hooks'

export default function NextApp({ Component, pageProps }) { 
 // ? Temp fix for FOUC on route change...  :(
  usePageTransitionFix()
// more app code...
}
```
_use-page-transition-fix.js hook..._
```jsx
import Router from 'next/router'
import { useEffect } from 'react'

export const OPACITY_EXIT_DURATION = 1

const routeChange = () => {
  const tempFix = () => {
    const elements = document.querySelectorAll('style[media="x"]')
    elements.forEach((elem) => elem.removeAttribute('media'))
    setTimeout(() => {
      elements.forEach((elem) => elem.remove())
    }, OPACITY_EXIT_DURATION * 1000)
  }
  tempFix()
}

export const usePageTransitionFix = () => {
  console.debug(
    'WARNING: Still using FOUC temp fix on route change.  Has the Next.js bug not been fixed?  See https://github.com/vercel/next.js/issues/17464',
  )
  useEffect(() => {
    Router.events.on('routeChangeComplete', routeChange)
    Router.events.on('routeChangeStart', routeChange)

    return () => {
      Router.events.off('routeChangeComplete', routeChange)
      Router.events.off('routeChangeStart', routeChange)
    }
  }, [])

  useEffect(() => {
    const asPath = Router.router?.asPath
    Router.router?.push(asPath)
    // ? Use replace() instead of push()?
  }, [])
}
```

**Comment by JoshuaHolloway** [comment]
> Check this: [#17464 (comment)](https://github.com/vercel/next.js/issues/17464#issuecomment-1080337949) it is the code from above

I can confirm that this temporary fix appears to have solved my problem with failing CSS-modules styling during page transitions using react-transition group when using 'npm run build' (next build).

These are my relavent dependencies.

  "dependencies": {
    "@headlessui/react": "^1.6.1",
    "@heroicons/react": "^1.0.6",
    "@mantine/core": "^4.1.4",
    "@mantine/hooks": "^4.1.4",
    "@mantine/next": "^4.1.4",
    "@react-google-maps/api": "^2.10.1",
    "@tailwindcss/forms": "^0.5.2",
    "next": "12.1.6",
    "react": "17.0.2",
    "react-dom": "17.0.2",
    "react-transition-group": "^4.4.2",
    "styled-components": "^5.3.3",
  },
  "devDependencies": {
    "autoprefixer": "^10.4.7",
    "postcss": "^8.4.14",
    "sass": "^1.50.0",
    "tailwindcss": "^3.0.24"
  }

**Comment by snatvb** [comment]
the first transition still breaks styles

**Comment by Simon-Renault** [comment]
@snatvb Same for me, did you find a way to fix it ?

**Comment by estorski** [comment]
Seems like this (somewhat critical) issue exists untouched for almost 2 years by now. If no action taken, it will probably stay that way for a few years more.

How can we put a higher priority on it?

**Comment by evantbyrne** [comment]
@estorski You can expedite it by submitting a pull request with a fix. :)

**Comment by wyattjoh** [comment]
Currently, the way that Next.js handles CSS during route changes is technically _correct_. Libraries like `framer-motion` retain the element render tree after they would have been removed traditionally (to animate the transition). Next.js removes the styles as soon as it replaces the page components during transitions, thereby causing the flashing of styles that you see today without using `styled-jsx` (which inlines styles in the tree rather than as a side-effect).

This is explained in more detail here https://github.com/vercel/next.js/issues/17464#issuecomment-751267740 and https://github.com/vercel/next.js/issues/17464#issuecomment-763739753.

We recognize that this an important use case and will update this issue as we investigate improvements to asset management.

**Comment by andrei-bancos** [comment]
Any fix that also works with dynamic routes..?

**Comment by estorski** [comment]
> Currently, the way that Next.js handles CSS during route changes is technically _correct_. Libraries like `framer-motion` retain the element render tree after they would have been removed traditionally (to animate the transition). Next.js removes the styles as soon as it replaces the page components during transitions, thereby causing the flashing of styles that you see today without using `styled-jsx` (which inlines styles in the tree rather than as a side-effect).
> 
> This is explained in more detail here [#17464 (comment)](https://github.com/vercel/next.js/issues/17464#issuecomment-751267740) and [#17464 (comment)](https://github.com/vercel/next.js/issues/17464#issuecomment-763739753).
> 
> We recognize that this an important use case and will update this issue as we investigate improvements to asset management.

Well, one possible solution could be - to DEregister the CSS file only after you made sure the current page markup does NOT have a single user of said styles. For that you could run something like `document.querySelector('.class1, .class2 ...')`, and if there's anything returned - delay the removal for a few seconds, and then repeat the procedure.

**Comment by snatvb** [comment]
@Simon-Renault no :(

**Comment by ThaoD5** [comment]
Coming with a non-solution to a problem that persists... 
If you're stuck; here is an approach, as ridiculous as it is; but keep in mind this is to be tested again when updating nextjs for obvious reasons

The function : 
```
  const savePageStyles = () => {
    const head = document.head
    const previousStylesFixes = head.querySelectorAll('[data-fix]')

    // Delete previously created fixes
    if (previousStylesFixes) {
      for (let i = 0; i < previousStylesFixes.length; i++) {
        head.removeChild(previousStylesFixes[i])
      }
    }

    // Get all the styles of the page
    const allStyleElems = head.querySelectorAll(
      'link[rel="stylesheet"], link[as="style"]'
    )
    // Get all the inline styles of the page, labelled by "data-n-href" ( defined by nextjs )
    const allInlineStylesElems = head.querySelectorAll('style[data-n-href]')

    // Create doubling links to css sheets that wont be removed unless we say so
    if (allStyleElems) {
      for (let i = 0; i < allStyleElems.length; i++) {
        if (allStyleElems[i].href) {
          const styles = document.createElement('link')
          styles.setAttribute('data-pt-fix', 'true')
          styles.setAttribute('rel', 'stylesheet')
          styles.setAttribute('href', allStyleElems[i].href)

          head.appendChild(styles)
        }
      }
    }

    // Now do the same with the inline styles
    const inlineStyles = document.createElement('style')
    inlineStyles.setAttribute('data-pt-fix', 'true')
    if (allInlineStylesElems) {
      for (let i = 0; i < allInlineStylesElems.length; i++) {
        inlineStyles.innerHTML += allInlineStylesElems[i].innerHTML
      }

      head.appendChild(inlineStyles)
    }
  }
```

When I call it : 

```
import { useRouter } from 'next/router'
...

const router = useRouter()
...

useEffect(() => {
  const handleDone = () => {
    // Arbitrary setTimeout, modify it at will
    setTimeout(() => {
      savePageStyles()
    }, 1000)
  }

  router.events.on('routeChangeComplete', handleDone)

  return () => {
    router.events.off('routeChangeComplete', handleDone)
  }
}, [router])
```

I wish this solution to be quickly unuseful.

**Comment by Mbiakop1** [comment]
> 

I looked at this issue and this seems to be happening because of the behaviour implemented in onCommit method.

**Comment by danieljamesross** [comment]
Getting the same issue on page load with next 12 and styled-components

**Comment by mmalomo** [comment]
Still happening with version `12.3.1`

**Comment by whiteand** [comment]
Still happening with version 13.0.0

**Comment by BenjaminIsMyName** [comment]
I'm moving my entire project to Tailwind because of this bug...

**Comment by antoine-lin** [comment]
Hi, I have the same issue using [vanilla-extract](https://vanilla-extract.style/) on version 13.0.0.

**Comment by simontaisne** [comment]
Same issue with `vanilla-extract` on `12.3.1`.

**Comment by jdlk07** [comment]
@danieljamesross I was having the same issue with styled-components on page load or page refresh. Just needed to create a `.babelrc` file in your root and add the following 

```
{
  "presets": ["next/babel"],
  "plugins": [["styled-components", { "ssr": true }]]
}
```

This was mentioned in the docs somewhere but I overlooked it. Not sure if it will work for you but it did for me.

**Comment by ArthurGoupil** [comment]
> 

This is working as a temporary fix.
To make the removal work I had to fix small typo
` const previousStylesFixes = head.querySelectorAll('[data-fix]')`
->
` const previousStylesFixes = head.querySelectorAll('[data-pt-fix]')`

**Comment by ArthurGoupil** [comment]
> Coming with a non-solution to a problem that persists... If you're stuck; here is an approach, as ridiculous as it is; but keep in mind this is to be tested again when updating nextjs for obvious reasons
> 
> The function :
> 
> ```
>   const savePageStyles = () => {
>     const head = document.head
>     const previousStylesFixes = head.querySelectorAll('[data-fix]')
> 
>     // Delete previously created fixes
>     if (previousStylesFixes) {
>       for (let i = 0; i < previousStylesFixes.length; i++) {
>         head.removeChild(previousStylesFixes[i])
>       }
>     }
> 
>     // Get all the styles of the page
>     const allStyleElems = head.querySelectorAll(
>       'link[rel="stylesheet"], link[as="style"]'
>     )
>     // Get all the inline styles of the page, labelled by "data-n-href" ( defined by nextjs )
>     const allInlineStylesElems = head.querySelectorAll('style[data-n-href]')
> 
>     // Create doubling links to css sheets that wont be removed unless we say so
>     if (allStyleElems) {
>       for (let i = 0; i < allStyleElems.length; i++) {
>         if (allStyleElems[i].href) {
>           const styles = document.createElement('link')
>           styles.setAttribute('data-pt-fix', 'true')
>           styles.setAttribute('rel', 'stylesheet')
>           styles.setAttribute('href', allStyleElems[i].href)
> 
>           head.appendChild(styles)
>         }
>       }
>     }
> 
>     // Now do the same with the inline styles
>     const inlineStyles = document.createElement('style')
>     inlineStyles.setAttribute('data-pt-fix', 'true')
>     if (allInlineStylesElems) {
>       for (let i = 0; i < allInlineStylesElems.length; i++) {
>         inlineStyles.innerHTML += allInlineStylesElems[i].innerHTML
>       }
> 
>       head.appendChild(inlineStyles)
>     }
>   }
> ```
> 
> When I call it :
> 
> ```
> import { useRouter } from 'next/router'
> ...
> 
> const router = useRouter()
> ...
> 
> // On router change, trigger that function
> useEffect(() => {
>   const handleDone = () => {
>     // Arbitrary setTimeout, modify it at will
>     setTimeout(() => {
>       savePageStyles()
>     }, 1000)
>   }
> 
>   router.events.on('routeChangeComplete', handleDone)
> 
>   return () => {
>     router.events.off('routeChangeComplete', handleDone)
>   }
> }, [router])
> 
> // Trigger the onMount function on the first page load so that your first page is also "fixed"
> // Note : I let you decide how this function is triggered because this depends a lot per stack
> const onMount = () => {
>   savePageStyles()
> }
> ```
> 
> I wish this non-solution to be quickly unuseful.

This is working as a temporary fix.
To make the removal work I had to fix small typo
`const previousStylesFixes = head.querySelectorAll('[data-fix]')`
->
`const previousStylesFixes = head.querySelectorAll('[data-pt-fix]')`

**Comment by wolfgangschoeffel** [comment]
> @JoshuaHolloway in __app.js_...
> 
> ```js
> import { usePageTransitionFix } from '@/hooks'
> 
> export default function NextApp({ Component, pageProps }) { 
>  // ? Temp fix for FOUC on route change...  :(
>   usePageTransitionFix()
> // more app code...
> }
> ```
> 
> _use-page-transition-fix.js hook..._
> 
> ```js
> import Router from 'next/router'
> import { useEffect } from 'react'
> 
> export const OPACITY_EXIT_DURATION = 1
> 
> const routeChange = () => {
>   const tempFix = () => {
>     const elements = document.querySelectorAll('style[media="x"]')
>     elements.forEach((elem) => elem.removeAttribute('media'))
>     setTimeout(() => {
>       elements.forEach((elem) => elem.remove())
>     }, OPACITY_EXIT_DURATION * 1000)
>   }
>   tempFix()
> }
> 
> export const usePageTransitionFix = () => {
>   console.debug(
>     'WARNING: Still using FOUC temp fix on route change.  Has the Next.js bug not been fixed?  See https://github.com/vercel/next.js/issues/17464',
>   )
>   useEffect(() => {
>     Router.events.on('routeChangeComplete', routeChange)
>     Router.events.on('routeChangeStart', routeChange)
> 
>     return () => {
>       Router.events.off('routeChangeComplete', routeChange)
>       Router.events.off('routeChangeStart', routeChange)
>     }
>   }, [])
> 
>   useEffect(() => {
>     const asPath = Router.router?.asPath
>     Router.router?.push(asPath)
>     // ? Use replace() instead of push()?
>   }, [])
> }
> ```

Make sure to clear the timeout when routeChange is called, otherwise when the user navigates back during the transition,  the styles of the current page will be removed.

Also when using query params like this:
```js
useEffect(() => {
    const pathname = Router.router?.pathname
    const query = Router.router?.query
    Router.router?.push({ pathname, query })
  }, [])
```
a 404 or 500 will redirect to /_error instead of displaying an error message on the current page. 

This is my modification. It can probably be improved. Just wanted to point out these two issues because I already saw them in the wild. 

```js
import Router from 'next/router'
import { useEffect } from 'react'

export const OPACITY_EXIT_DURATION = 1

export const useTransitionFix = () => {
  useEffect(() => {
    let timeout
    const routeChange = () => {
      clearTimeout(timeout)
      const elements = document.querySelectorAll('style[media="x"]')
      elements.forEach((elem) => elem.removeAttribute('media'))
      timeout = setTimeout(() => {
        elements.forEach((elem) => elem.remove())
      }, OPACITY_EXIT_DURATION * 1000)
    }

    Router.events.on('routeChangeComplete', routeChange)
    Router.events.on('routeChangeStart', routeChange)

    return () => {
      Router.events.off('routeChangeComplete', routeChange)
      Router.events.off('routeChangeStart', routeChange)
      clearTimeout(timeout)
    }
  }, [])

  useEffect(() => {
    const pathname = Router.router?.pathname
    const query = Router.router?.query
    if (pathname === '/_error') {
      return
    }
    Router.router?.push({ pathname, query })
  }, [])

  return null
}

**Comment by shaianest** [comment]
had the same issue with work around above but my css module is in _app and the hack has flicker even without timeout
my next version is 12.3.4 and i'm having multiple other apps that work as expected.

**Comment by Kethatril** [comment]
Using this fix https://github.com/moxystudio/next-with-moxy/blob/master/www/app/use-fouc-fix.js from @satazor has fixed any issues for us (calling `useFoucFix()` in `_app.jsx`)

**Comment by KingBael09** [comment]
I've got a workaround without using the fix by @satazor!

By importing component dynamically (next/dynamic).....the problem of CSS-disabling goes away.

```Javascript
// index.js

import dynamic from "next/dynamic";

const Component1 = dynamic(() => import("@path/component1"));

const Component2 = dynamic(() => import("@path/component2"));

export default function Home() {
  return (
    <>
      <Component1/>
      <Component2/>
    </>
  );
}
```

However Page should contain only components & each component has its own CSS Module.

This workaround works in my case Next.js 13.1.2 and framer-motion 8.5.2

**Comment by lorenzomigliorero** [comment]
Apparently, there's no way to avoid CSS splitting (even for < 1kb file!).
As a rude solution, I merged all the chunked CSS into the global stylesheet (assuming that it starts with the `.html` selector):

```js
const path = require('path');
const fs = require('fs');

const CSS_PATH = path.resolve(__dirname, './.next/static/css');

const CSS_FILENAMES = fs.readdirSync(CSS_PATH);

const GLOBAL_CSS_FILENAME = CSS_FILENAMES.find((file) => {
  const FILE_PATH = path.join(CSS_PATH, file);
  const data = fs.readFileSync(FILE_PATH, 'utf8');
  return data.startsWith('html');
});

const GLOBAL_CSS_PATH = path.join(CSS_PATH, GLOBAL_CSS_FILENAME);

CSS_FILENAMES.filter((name) => name !== GLOBAL_CSS_FILENAME).forEach((file) => {
  const FILE_PATH = path.join(CSS_PATH, file);
  const data = fs.readFileSync(FILE_PATH, 'utf8');

  fs.appendFileSync(GLOBAL_CSS_PATH, data);
  fs.truncate(FILE_PATH, 0, () => console.log('done'));
});
```

**Comment by claus** [comment]
Oh, this thread is still alive.

I have an update on the hook i posted here almost two years ago: https://github.com/vercel/next.js/issues/17464#issuecomment-796430107

The new version below solves a couple of things:
- All dynamic page styles except the latest two are now being removed as new page styles come in (this could be improved further, but is good enough for our use cases).
- We ran into weird behavior with the original hook where removing `media="x"` wasn't fast enough to prevent the browser from noticing and recalculating styles and running layout, but not fast enough for it to run paint and composition, resulting in scroll position jumping up without anything in the page actually changing. Took me a while to figure this one out.

Just call this hook in your `_app` and you should be all set.

Don't forget to set `scroll` to `false` in your `<Link>`s and `router.push`s.

```js
import * as React from 'react';

export const useNextCssRemovalPrevention = () => {
    React.useEffect(() => {
        // Remove data-n-p attribute from all link nodes.
        // This prevents Next.js from removing server rendered stylesheets.
        document.querySelectorAll('head > link[data-n-p]').forEach(linkNode => {
            linkNode.removeAttribute('data-n-p');
        });

        const mutationHandler = (mutations: MutationRecord[]) => {
            mutations.forEach(({ target, addedNodes }: MutationRecord) => {
                if (target.nodeName === 'HEAD') {
                    // Add data-n-href-perm attribute to all style nodes with attribute data-n-href,
                    // and remove data-n-href and media attributes from those nodes.
                    // This prevents Next.js from removing or disabling dynamic stylesheets.
                    addedNodes.forEach(node => {
                        const el = node as Element;
                        if (el.nodeName === 'STYLE' && el.hasAttribute('data-n-href')) {
                            const href = el.getAttribute('data-n-href');
                            if (href) {
                                el.setAttribute('data-n-href-perm', href);
                                el.removeAttribute('data-n-href');
                                el.removeAttribute('media');
                            }
                        }
                    });

                    // Remove all stylesheets that we don't need anymore
                    // (all except the two that were most recently added).
                    const styleNodes = document.querySelectorAll('head > style[data-n-href-perm]');
                    const requiredHrefs = new Set<string>();
                    for (let i = styleNodes.length - 1; i >= 0; i--) {
                        const el = styleNodes[i];
                        if (requiredHrefs.size < 2) {
                            const href = el.getAttribute('data-n-href-perm');
                            if (href) {
                                if (requiredHrefs.has(href)) {
                                    el.parentNode!.removeChild(el);
                                } else {
                                    requiredHrefs.add(href);
                                }
                            }
                        } else {
                            el.parentNode!.removeChild(el);
                        }
                    }
                }
            });
        };

        // Observe changes to the head element and its descendents.
        const observer = new MutationObserver(mutationHandler);
        observer.observe(document.head, { childList: true, subtree: true });

        return () => {
            // Disconnect the observer when the component unmounts.
            observer.disconnect();
        };
    }, []);
};

```

**Comment by claus** [comment]
For the Next team (cc @wyattjoh)

To help mitigate this issue, maybe you could add a next.config.js flag that, if enabled, lets app developers opt out of the current behavior.

Some ideas of what this could do:

- opt out of the removal of styles altogether
- instead of removing styles for old pages right away, keep styles for the previous page and only remove those that are older than that
- a hook or API that provides a cleanup function that users can call when their page transition is done, which sets media="x" on old style elements, and does whatever else has to be done

**Comment by DangerousJack** [comment]
> @danieljamesross I was having the same issue with styled-components on page load or page refresh. Just needed to create a `.babelrc` file in your root and add the following
> 
> ```
> {
>   "presets": ["next/babel"],
>   "plugins": [["styled-components", { "ssr": true }]]
> }
> ```
> 
> This was mentioned in the docs somewhere but I overlooked it. Not sure if it will work for you but it did for me.

this worked for me. Just had to install babel-styled-components, create a babel config and paste in the setup to use the styled components plugin. So far so good! Thanks!

**Comment by panteliselef** [comment]
@DangerousJack I think this works only if you are using babel and not SWC, which is not the case for most people as Next.js is using SWC by default these days

**Comment by mxydl2009** [comment]
with "next": "13.1.6" and "framer-motion": "^10.10.0", this page transition problem is still alive, and through above solutions, can solve the problem in my project. thank you!

**Comment by claus** [comment]
> I have an update on the hook i posted here almost two years ago: [#17464 (comment)](https://github.com/vercel/next.js/issues/17464#issuecomment-796430107)

Just a heads up, that hook i posted above assumes that one page loads exactly one stylesheet. That's not always the case. In some configurations, pages load more than one stylesheet, breaking my hook. I'll have a fix for that soon.

**Comment by claus** [comment]
Ok, probably my last post on this issue :)

We published a PageTransition suite of components, that includes the `useNextCssRemovalPrevention` hook I've been posting about here (https://github.com/vercel/next.js/issues/17464#issuecomment-796430107 and https://github.com/vercel/next.js/issues/17464#issuecomment-1447335147). It prevents Next.js from removing page styles too early, but removes them eventually. It addresses all problems I'm currently aware of.

**Docs**
https://hauscore.xyz/components-nextjs/page-transition
https://hauscore.xyz/components-nextjs/page-transition#usenextcssremovalprevention

**Demo**
https://next-starter-ts.hauscore.xyz (ts)
https://next-starter.hauscore.xyz (js)

**Comment by nicholasruggeri** [comment]
@claus <3

**Comment by baptistebriel** [comment]
Thank you @claus 🙏🏻

**Comment by ArianHamdi** [comment]
> Ok, probably my last post on this issue :)
> 
> We published a PageTransition suite of components, that includes the `useNextCssRemovalPrevention` hook I've been posting about here ([#17464 (comment)](https://github.com/vercel/next.js/issues/17464#issuecomment-796430107) and [#17464 (comment)](https://github.com/vercel/next.js/issues/17464#issuecomment-1447335147)). It prevents Next.js from removing page styles too early, but removes unused styles eventually. It addresses all problems I'm currently aware of.
> 
> `npm i @madeinhaus/nextjs-page-transition`
> 
> `import { useNextCssRemovalPrevention } from '@madeinhaus/nextjs-page-transition'`
> 
> **Source** https://github.com/MadeInHaus/haus-core/blob/main/packages/nextjs-page-transition/src/helpers/useNextCssRemovalPrevention.ts
> 
> **Docs** https://hauscore.xyz/components-nextjs/page-transition https://hauscore.xyz/components-nextjs/page-transition#usenextcssremovalprevention
> 
> **Demo** https://next-starter-ts.hauscore.xyz (ts) https://next-starter.hauscore.xyz (js)

Love U
